Fuse row-wise sharded linear matmul to increase perf.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78598
* #78477
* __->__ #78449



Instead of looping through and performing a matmul separately, we can
just perform a single matmul to ensure we launch a single cuda kernel for this
operation.

Differential Revision: [D36743354](https://our.internmc.facebook.com/intern/diff/D36743354/)

Differential Revision: [D36743354](https://our.internmc.facebook.com/intern/diff/D36743354)

pritamdamania87 committed:
Fuse row-wise sharded linear matmul to increase perf.

Instead of looping through and performing a matmul separately, we can
just perform a single matmul to ensure we launch a single cuda kernel for this
operation.

Differential Revision: [D36743354](https://our.internmc.facebook.com/intern/diff/D36743354/)

[ghstack-poisoned]

pritamdamania87 committed:
Update on "Fuse row-wise sharded linear matmul to increase perf."


Instead of looping through and performing a matmul separately, we can
just perform a single matmul to ensure we launch a single cuda kernel for this
operation.

Differential Revision: [D36743354](https://our.internmc.facebook.com/intern/diff/D36743354/)

[ghstack-poisoned]

pritamdamania87 committed:
Update on "Fuse row-wise sharded linear matmul to increase perf."




Instead of looping through and performing a matmul separately, we can
just perform a single matmul to ensure we launch a single cuda kernel for this
operation.

Differential Revision: [D36743354](https://our.internmc.facebook.com/intern/diff/D36743354/)

Differential Revision: [D36743354](https://our.internmc.facebook.com/intern/diff/D36743354)

[ghstack-poisoned]

vadimkantorov commented on torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/linear.py:
```diff
@@ -357,8 +353,9 @@ def forward(ctx, world_size, start_pos, chunk_size, weight, pg, bias):
     @staticmethod
     def backward(ctx, grad_output):
         results = []
+        grad_output_cloned = grad_output.clone()
         for idx in range(ctx.world_size):
-            results.append(grad_output.clone())
+            results.append(grad_output_cloned)
```
Could this be reformulated as `results = [grad_output.clone()] * ctx.world_size`?

aazzolini commented on torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/linear.py:
```diff
@@ -323,21 +323,17 @@ def _handle_row_wise_sharding_sharded_tensor(
-        results.append(
-            tensor.matmul(local_shard_t) + _BiasTensorPartial.apply(world_size, bias)
-        )
+    result = local_shard.matmul(local_shard_t) + _BiasTensorPartial.apply(world_size, bias)
```
Would addmm allow to fuse the bias addition?

pritamdamania87 committed:
Update on "Fuse row-wise sharded linear matmul to increase perf."




Instead of looping through and performing a matmul separately, we can
just perform a single matmul to ensure we launch a single cuda kernel for this
operation.

Differential Revision: [D36743354](https://our.internmc.facebook.com/intern/diff/D36743354/)

Differential Revision: [D36743354](https://our.internmc.facebook.com/intern/diff/D36743354)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/linear.py:
```diff
@@ -323,21 +323,17 @@ def _handle_row_wise_sharding_sharded_tensor(
     """
-    results = []
-    local_shard = input.local_shards()[0].tensor
+    local_input = input.local_shards()[0].tensor
```
nit: input.local_tensor() to trigger the check of single local shard.

wanchaol reviewed (approved):
Looks great!

pritamdamania87 committed:
Update on "Fuse row-wise sharded linear matmul to increase perf."




Instead of looping through and performing a matmul separately, we can
just perform a single matmul to ensure we launch a single cuda kernel for this
operation.

Differential Revision: [D36743354](https://our.internmc.facebook.com/intern/diff/D36743354/)

Differential Revision: [D36743354](https://our.internmc.facebook.com/intern/diff/D36743354)

[ghstack-poisoned]

pritamdamania87 commented:
@pytorchbot merge this please
